### DISCUSSION TITLE
Next's `build` includes a peerDependency from an imported module

### DISCUSSION THREAD

**Discussion by matiishyn** [vercel/next.js#25060] [Help]
### What version of Next.js are you using?

latest

### What version of Node.js are you using?

14.15.1

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

locally

### Describe the Bug

I have a lerna monorepo with such packages ([link to Github](https://github.com/matiishyn/next-peer-deps)):
- next-app: TS next.js app, imports `shared` module
- shared: TS package with some shared React components

Now,
- `next-app` has `@apollo/client` in its dependencies and imports a component from `shared`
- `shared` has `@apollo/client` in its peerDependencies (can be also in devDependencies) and uses Apollo to fetch some data

After running bundle-analyze I can see that `@apollo/client` was included 2 times:
1. inside node_modules, which seems correct
2. inside shared/node_modules which is not quite correct, since it's a peerDependency

![image](https://user-images.githubusercontent.com/3661065/118001675-9fd11480-b34f-11eb-8b18-ec3d5f78b2e6.png)

### Expected Behavior

`@apollo/client` should be included only once

### To Reproduce

1. clone repo https://github.com/matiishyn/next-peer-deps
2. `npm i`
3. `npm run build-shared`
4. `npm run build-analyze` - this will display the Bundle Analyzer as on my screenshot

**Comment by matiishyn** [comment]
I've found something that seems to fix the issue but not sure if that's a correct way
https://github.com/matiishyn/next-peer-deps/blob/master/packages/next-app/next.config.js#L17

**Comment by timneutkens** [comment]
The reason is most likely that you have 2 different versions of the same dependency installed, one in `./node_modules` and one in `./shared/node_modules`, you'll likely want to get rid of one of the deps.

**Reply by matiishyn** [reply]
I have the exact same version for both packages
![image](https://user-images.githubusercontent.com/3661065/118021290-c26d2880-b363-11eb-83d9-edc4396e1c4f.png)

When I build `shared` alone then it does not include the `@apollo/client`
![image](https://user-images.githubusercontent.com/3661065/118022016-8be3dd80-b364-11eb-9ae8-7b83f3f33070.png)

But when I build my next-app it includes both `node_modules/@apollo/client` and `shared/node_modules/@apollo/client`.

------
> The reason is most likely that you have 2 different versions of the same dependency installed, one in ./node_modules and one in ./shared/node_modules, you'll likely want to get rid of one of the deps.

NPM7 [does automatically install peerDependencies ](https://github.blog/2020-10-13-presenting-v7-0-0-of-the-npm-cli/#user-content-exciting-new-features) but I don't think that they should be included in the module that uses the `shared` module. Maybe that can be somehow handled by Next.js (or its Webpack).

@timneutkens can you please advice what can be done here?

**Reply by timneutkens** [reply]
You'll probably want to hoist them to a shared top-level directory because currently you have 2 different versions installed as it's likely not using a yarn workspaces type solution from the looks of it.
